include/oneapi/dpl/pstl/glue_algorithm_ranges_impl.h

```diff
@@ -24,10 +24,195 @@
 #    include "hetero/algorithm_impl_hetero.h" //TODO: for __brick_copy
 #endif
 
+#if _ONEDPL___cplusplus >= 202002L
+#    include <ranges>
+#    include "algorithm_ranges_impl.h"
+#endif
+
 namespace oneapi
 {
 namespace dpl
 {
+
+#if _ONEDPL___cplusplus >= 202002L
+namespace ranges
+{
+
+// [alg.foreach]
+
+struct for_each_fn
+{
+    template<typename _ExecutionPolicy, typename _R, typename _Proj, typename _Fun,
+             oneapi::dpl::__internal::__enable_if_execution_policy<_ExecutionPolicy, int> = 0>
+    constexpr auto
+    operator()(_ExecutionPolicy&& __exec, _R&& __r, _Fun __f, _Proj __proj) const
+    {
+        const auto __dispatch_tag = oneapi::dpl::__internal::__select_backend(__exec, __r.begin());
```

Comment:
Continuing the dialog in the previous PR:

>> Does it make sense to support ranges directly in `__select_backend`?
>
> At the current moment I don't see value from that... (If I don't miss anything)

Perhaps it can be considered a syntax sugar. On the other hand, the implementation can check the range concepts instead of iterator traits.

Reply:
to check the range concepts - it might be a value here.
An range overload for __select_backend:
1) We can make the change in the future when we write concept support in oneDPL.
2) We can make it now just for "placeholder" perspective and don't change the code of  the first dispatch level.
Which way is preferable now?

Reply:
I have no strong preference; suggesting you to decide.

Reply:
I am afraid, that variant 2 is not implementable right now due to we have got ambiguity w/o concepts, because now we have a general parameter pack like this
`__select_backend(policy, _IteratorTypes&&...)`

Reply:
Actually, after the concepts for introduced Range API have been added, the situation is becoming simpler  - we need only policy type check! Due to just random access ranges are allowed we don't need an access type (random/forward) at all.
`const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);`
So, `__ranges::__select_backend(exec) ` can be implemented via 4 overloads for 4 host policies, that's all.

Reply:
Moreover, it seems the `oneapi::dpl::__internal::__select_backend` work with "policy only" in proper way:
https://github.com/oneapi-src/oneDPL/blob/main/include/oneapi/dpl/pstl/execution_impl.h#L91
So, we dont need introduce a new one,  like `oneapi::dpl::__ranges::__select_backend`.

Reply:
For the latest observation, the only thing that I am worried about is how to make sure that we "intentionally miss optional parameters" (like `IteratorTypes` or `RangesType`) compared to "forgot to pass optional parameters"?

I was considering to require at least one `IteratorType` parameter for the C++17 algorithms for more robust API. Maybe we should have some mandatory parameter for ranges as well... How about `__range_alg_t` tag? So, it would be something like:

```cpp
auto __select_backend(const parallel_policy&, range_alg_t)
```

With that tag we explicitly state our intent and make the "intentionally missing arguments" aspect unquestionable.

Today, we have `random_access_<iterator|range>`. It's not well-extensible in the future but I can live with that for now.

Reply:
I would keep `const auto __dispatch_tag = oneapi::dpl::__ranges::__select_backend(__exec);`